[DX-1781] Release notes for 5.3.9 (#5826)
TykTechnologies/tyk-docs: tyk-docs/content/developer-support/release-notes/gateway.md

## 5.3 Release Notes

### 5.3.9 Release Notes

#### Release Date xxx

#### Release Highlights

> This release focuses mainly on bug fixes. For a comprehensive list of changes, please refer to the detailed [changelog]({{< ref "#Changelog-v5.3.9">}}) below.

Reviewer: ```suggestion
This release contains a bug fix. For a comprehensive list of changes, please refer to the detailed [changelog]({{< ref "#Changelog-v5.3.9">}}) below.
```

Here it is important to explain the benefit of each changelog item. As mentioned by James in a previous Slack message (https://tyktech.slack.com/archives/C044R3ZTN6L/p1686812207060839?thread_ts=1686762128.651249&cid=C044R3ZTN6L):
"...it is important to document the customer impact for the work delivered, so we can share it with prospects/install base. For example:
"New Chart delivers x and y benefit to a and b customer use cases. The business impact for them will be this and that" -->
##### Fixed

<ul>
<li>
<details>
<summary>Resolved analytics data trimming with custom plugins</summary>

Resolved an issue where analytics data was incorrectly trimmed when a custom response plugin (coprocess) modified the response body to exceed the upstream content length.
</details>
</li>
<li>
<details>
<summary>Skip loading API security feature adjustments for Cloud users</summary>

Fixed a bug that prevented management gateways from loading plugin bundles due to a security feature introduced in v5.3.1.
</details>
</li>
<li>
<details>

> <summary>Debug log accuracy restored for middlewares</summary>

Reviewer 2: The issue here was that the system log stopped differentiating between different middleware, making it very difficult to debug issues.

> Addressed an issue where shared loggers caused misreported debug logs for middlewares. Loggers are now explicitly copied to ensure accurate and distinct logging.

Reviewer 2: I think the first sentence is fine, but for the second I think we just need to make it clear that the log entries will now correctly identify the source of the log entry (i.e. which middleware generated the log).

</details>
</li>
<li>
<details>

> <summary>Otel E2E Tests Issue After Upgrading to v1.29</summary>

Reviewer 2: This feels very much like an internal fix

Upgrading the OpenTelemetry SDK to v1.29 caused otelhttp v0.54.0 to break the CI pipelines. As a temporary fix, we reverted to otelhttp v0.49.0 while investigating the root cause of the issue.
</details>
</li>
<li>
<details>

> <summary>Resolved API routing issue with trailing slashes and overlapping listen paths</summary>

Reviewer 2: Perhaps here the summary should just state that we've fixed some inconsistencies in the routing algorithm that led to incorrect API URL matching.

Fixed a routing issue that caused incorrect API matching when dealing with APIs that lacked a trailing slash, used custom domains, or had similar listen path patterns. Previously, the router prioritized APIs with longer subdomains and shorter listen paths, leading to incorrect matches when listen paths shared prefixes. This fix ensures accurate API matching, even when subdomains and listen paths overlap.
</details>
</li>
<li>
<details>

> <summary>SSE streaming reliability enhanced</summary>

Reviewer 2: I think here we can be explicit - we've not enhanced SSE streaming reliability, we've fixed a bug where the Gateway would panic if SSE was used with OTel

> Fixed an issue where upstream server-sent events (SSE) were not sent when OpenTelemetry was enabled.Resolved a gateway panic triggered when detailed recording was enabled and upstream sent SSE.

Reviewer 2: From my reading of the ticket, the issue was that the Gateway would panic (crash) - probably worth checking with @jeffy-mathew, as the release note on the ticket does mention two fixes, but the ticket description only mentions the panic.

</details>
</li>
<li>
<details>

> <summary>OAuth 2.0 token issuance now supported in Emergency mode</summary>

Reviewer 2: I don't see "Emergency Mode" anywhere in the docs - so would recommend something like:

```suggestion
<summary>Fixed an issue where OAuth 2.0 access tokens would not be issued if the data plane was disconnected from the control plane</summary>
```

> OAuth 2.0 tokens can now be issued even when data plane gateways are in emergency mode. This is achieved by saving OAuth clients locally when they are pulled from RPC.

Reviewer 2: ```suggestion
OAuth 2.0 access tokens can now be issued even when data plane gateways are disconnected from the control plane. This is achieved by saving OAuth clients locally within the data plane when they are pulled from RPC.
```

</details>
</li>
<li>
<details>

> <summary>Request size limit Middleware no longer breaks GET or DELETE requests</summary>

Reviewer 2: "Breaks" is a bit vague - I would suggest something like:

```suggestion
<summary>Request size limit middleware would block any request without a payload (for example GET, DELETE)</summary>
```

> Resolved a problem where the request size limit middleware was incorrectly breaking GET and DELETE requests.

Reviewer 2: I think we can explain here that the middleware would expect to see a request body (payload) and fail validation if none was found, effectively blocking GET and DELETE methods.

</details>
</li>
<li>
<details>
<summary>Accurate debug logging restored for middleware</summary>

Addressed an issue where shared loggers caused debug logs to misidentify the middleware source, complicating debugging. Log entries now correctly indicate which middleware generated the log, ensuring clearer and more reliable diagnostics
</details>
</li>
<li>
<details>
<summary>API Keys remain active after all linked partitioned policies are deleted</summary>

Resolved an issue where API access keys remained valid even if all associated policies were deleted. The Gateway now attempts to apply all linked policies to the key when it is presented with a request. Warning logs are generated if any policies cannot be applied (for example, if they are missing). If no linked policy can be applied, the Gateway will reject the key to ensure no unauthorized access.
</details>
</li>
<li>
<details>

> <summary>Resolved API routing issue with trailing slashes and overlapping listen paths</summary>

Reviewer: Does this bug fix specific to `5.3.9`? I cannot find the same in `5.7.1`

Author: It was in 5.7.0 so it only need to be 5.3.9
